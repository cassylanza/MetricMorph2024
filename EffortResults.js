import React, { Component } from 'react'
import { TouchableOpacity, Text, View, StyleSheet, ScrollView } from 'react-native'
import { Actions } from 'react-native-router-flux'


// formats the calculated time into hours, minutes, and seconds
// returns default of "hh:mm:ss" if input is invalid (i.e. not a number
// or outside of the specified range)
export function formatResult(timeInSeconds) {

  let maxNumber = 362399;

  if (isNaN(timeInSeconds) || typeof timeInSeconds !== 'number') {
      return "hh:mm:ss";
  }

  if(timeInSeconds < 0 || timeInSeconds > maxNumber) {
    return "hh:mm:ss";
  }

  let hours = parseInt((Number(timeInSeconds))/3600);
  let minutes = parseInt(Number(timeInSeconds)/60)-(hours*60);
  let seconds = (Number(timeInSeconds)%60);
  let formattedSplit = timeInSeconds;


  if (timeInSeconds < 60) {
    formattedSplit = seconds.toFixed(2);
  }
  else if (timeInSeconds < 60*60) {

    if (seconds < 10) {
      formattedSplit = minutes + ":0" + seconds.toFixed(2);
    } else {
      formattedSplit = minutes + ":" + seconds.toFixed(2);
    }
  }
  else {

    formattedSplit = hours;

    if (minutes < 10) {
      formattedSplit = formattedSplit + ":0" + minutes;
    } else {
      formattedSplit = formattedSplit + ":" + minutes;
    }

    if (seconds < 10) {
      formattedSplit = formattedSplit + ":0" + seconds.toFixed(2);
    } else {
      formattedSplit = formattedSplit + ":" + seconds.toFixed(2);
    }
  }
  return formattedSplit;
}


// main display component
// calculates a pace corresponding to an effort percentage
// converts given performance to 3200 performance and divides by given % effort
// returns main calculation with table underneath
const EffortResults = ({race, effortPercent, hours, minutes, seconds}) => {

  let basisDistance = 3200; // effort relies on 3200 pace

  let conversionExponent = 1.06; // value of exponent in Riegal's conversion formula

  let currentRaceTimeVar = Number(hours)*3600 + Number(minutes)*60 + Number(seconds);

  let mySplit = Number(currentRaceTimeVar) * (Number(basisDistance)/(Number(race))) ** (conversionExponent);

  const events = [
      {
        id: 0,
        effort: '65%',
        time: formatResult(Number(mySplit/2)/(Number(.65))),
      },
      {
        id: 1,
        effort: '70%',
        time: formatResult(Number(mySplit/2)/(Number(.70))),
      },
      {
        id: 2,
        effort: '75%',
        time: formatResult(Number(mySplit/2)/(Number(.75))),
      },
      {
        id: 3,
        effort: '80%',
        time: formatResult(Number(mySplit/2)/(Number(.8)))
      },
      {
        id: 4,
        effort: '85%',
        time: formatResult(Number(mySplit/2)/(Number(.85)))
      },
      {
        id: 5,
        effort: '90%',
        time: formatResult(Number(mySplit/2)/(Number(.9)))
      },
      {
        id: 6,
        effort: '95%',
        time: formatResult(Number(mySplit/2)/(Number(.95)))
      },
      {
        id: 7,
        effort: '96%',
        time: formatResult(Number(mySplit/2)/(Number(.96)))
      },
      {
        id: 8,
        effort: '97%',
        time: formatResult(Number(mySplit/2)/(Number(.97)))
      },
      {
        id: 9,
        effort: '98%',
        time: formatResult(Number(mySplit/2)/(Number(.98)))
      },
      {
        id: 10,
        effort: '99%',
        time: formatResult(Number(mySplit/2)/(Number(.99)))
      },
      {
        id: 11,
        effort: '100%',
        time: formatResult(Number(mySplit/2)/(Number(1)))
      },
      {
        id: 12,
        effort: '105%',
        time: formatResult(Number(mySplit/2)/(Number(1.05)))
      },
      {
        id: 13,
        effort: '110%',
        time: formatResult(Number(mySplit/2)/(Number(1.1)))
      },
      {
        id: 14,
        effort: '115%',
        time: formatResult(Number(mySplit/2)/(Number(1.15)))
      },
      {
        id: 15,
        effort: '120%',
        time: formatResult(Number(mySplit/2)/(Number(1.2)))
      }
    ];

    return (
      // displays main converion result followed by a table of
      // common effort percentages
      <View>
        <Text style = {styles.mainResult}> { formatResult((mySplit/(Number(2)))/(Number(effortPercent)/(Number(100)))) } </Text>
        <Text style = {{textAlign: 'center', marginBottom: 30}}> min/mile for { effortPercent }% </Text>
        <ScrollView>
          {events.map((item, index) => (
              <View key={item.id} style={styles.container}>
                <Text style={styles.text}>{item.effort}</Text>
                <Text style={styles.text}>{item.time}</Text>
              </View>
            ))}
        </ScrollView>
      </View>
    )
}
export default EffortResults

const styles = StyleSheet.create ({
  container: {
      padding: 10,
      marginTop: 3,
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'flex-end',
      borderWidth: 1
   },
   text: {
      color: '#4f603c'
   },
   mainResult: {
     width: 250,
     height: 75,
     textAlign: 'center',
     padding: 18,
     fontSize: 30,
     backgroundColor: 'white',
     marginTop: 60,
     marginBottom: 10,
     marginLeft: 60
   }
})
