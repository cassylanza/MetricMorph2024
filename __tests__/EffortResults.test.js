import React from 'react';
import { render, screen, fireEvent, getByText, toBe, queryAllByText, getByPlaceholderText } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';
import { Actions } from 'react-native-router-flux'; // Mocking react-native-router-flux
import '@testing-library/jest-dom';
import EffortResults from '../EffortResults.js';
import { formatResult } from '../EffortResults.js';

jest.mock('react-native-router-flux', () => ({
  Actions: {
    swimmenu1: jest.fn(),
    trackmenu1: jest.fn(),
    convperfinfoentry: jest.fn(),
    conversionresults: jest.fn(),
    splitsinfoentry: jest.fn(),
    splitsresults: jest.fn(),
    effortinfoentry: jest.fn(),
    effortresults: jest.fn(),
  },
}));

describe('EffortResults', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<EffortResults />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('EffortResults component', () => {
  it('displays the main result time correctly', () => {
    const race = 1600;
    const hours = 0;
    const minutes = 6;
    const seconds = 3;
    const effortPercent = 80;

    const { getByText } = render(
      <EffortResults race={race} hours={hours} minutes={minutes} seconds={seconds} effortPercent={effortPercent} />
    );

    const currentRaceTimeVar = hours*3600 + minutes*60 + seconds;
    const mySplit = Number(currentRaceTimeVar) * (Number(3200)/(Number(race))) ** (1.06);
    const expectedMainResult = formatResult((mySplit/(Number(2)))/(Number(effortPercent)/(Number(100))));
    const elements = screen.queryAllByText(expectedMainResult);

    expect(elements.length).toBeGreaterThan(0);
  });
});

describe('EffortResults component', () => {
it('displays the effort clarifier correctly', () => {
  const { getByText } = render(
    <EffortResults race={400} hours={0} minutes={5} seconds={0} effortPercent={80} />
  );

  const effortClarifier = getByText(' min/mile for 80% ');
  expect(effortClarifier).toBeTruthy();
});
});

  it('table appears', () => {
    const { getByText } = render(<EffortResults />);
    expect(getByText("65%")).toBeTruthy();
  });

  describe('formatResult function', () => {
    it('returns "hh:mm:ss" for invalid input', () => {
      expect(formatResult(undefined)).toBe('hh:mm:ss');
      expect(formatResult(null)).toBe('hh:mm:ss');
      expect(formatResult('abc')).toBe('hh:mm:ss');
    });

    it('returns "hh:mm:ss" if timeInSeconds is out of range', () => {
      expect(formatResult(-1)).toBe('hh:mm:ss');
      expect(formatResult(362400)).toBe('hh:mm:ss');
    });

    it('correctly formats time less than 60 seconds', () => {
      expect(formatResult(30)).toBe('30.00');
      expect(formatResult(59.99)).toBe('59.99');
    });

    it('correctly formats time between 60 seconds and 1 hour', () => {
      expect(formatResult(65)).toBe('1:05.00');
      expect(formatResult(3600)).toBe('1:00:00.00');
    });

    it('correctly formats time greater than 1 hour', () => {
      expect(formatResult(3665)).toBe('1:01:05.00');
      expect(formatResult(3666)).toBe('1:01:06.00');
      expect(formatResult(4200)).toBe('1:10:00.00');
      expect(formatResult(4210.23)).toBe('1:10:10.23')
      expect(formatResult(7200)).toBe('2:00:00.00');
    });
  });




})
