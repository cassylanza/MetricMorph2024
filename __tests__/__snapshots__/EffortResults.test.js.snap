// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EffortResults renders correctly 1`] = `
<View>
  <Text
    style={
      {
        "backgroundColor": "white",
        "fontSize": 30,
        "height": 75,
        "marginBottom": 10,
        "marginLeft": 60,
        "marginTop": 60,
        "padding": 18,
        "textAlign": "center",
        "width": 250,
      }
    }
  >
     
    hh:mm:ss
     
  </Text>
  <Text
    style={
      {
        "marginBottom": 30,
        "textAlign": "center",
      }
    }
  >
     min/mile for 
    % 
  </Text>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          65%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          70%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          75%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          80%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          85%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          90%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          95%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          96%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          97%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          98%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          99%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          100%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          105%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          110%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          115%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          120%
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
